### PR TITLE
[tests|tokenizers] Refactoring pipelines test backbone - Small tokenizers improvements - General tests speedups

### PR DESCRIPTION
**Pull request by thomwolf**
# What does this PR do?

This PR refactor the pipeline tests to split them in smaller parts more easy to iterate on.

There is now:
- one common backbone for testing pipelines in `test_pipeline_common.py` with two Mixin that can be used depending on the test: `CustomInputPipelineCommonMixin` and `MonoInputPipelineCommonMixin`. The later provide a standard `_test_pipeline(nlp: Pipeline)` method while the former require to write a custom test pipeline method. 
- one test file per specific pipeline inheriting fro the above backbone.

Small fixes:
- the special token ids can now be set in the tokenizers
- added `convert_tokens_to_string(List[str]) -> str` to the Fast Tokenizers
- added a `tokenizer.vocab` property in Fast Tokenizers (alias to `tokenizer.get_vocab()`)
- `tokenizer.decode()` now accept also PyTorch, TensorFlow, Numpy Tensors/arrays as input
- gathered a few docstring in the parent class for tokenizers
- also fix the Dialog Pipeline #5516

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to the it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

**Comment by thomwolf** [comment]
Tagging a few people working on the pipelines and @sshleifer because I've added additional imports of TF Bart in `modeling_tf_auto` to make the pipelines happy.

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -175,6 +175,23 @@ class TokenSpan(NamedTuple):
     end: int
 
 
+def to_py_obj(obj):
+    """
+    Convert a TensorFlow tensor, PyTorch tensor, Numpy array or python list
+    in a python list.
```
Comment (sshleifer) [inline src/transformers/tokenization_utils_base.py:181]:
```suggestion
    to a python list.
```

maybe also type hint -> List

**Review by sshleifer — APPROVED** [review]

File: src/transformers/pipelines.py [hunk 2]
```diff
@@ -2749,6 +2748,8 @@ def pipeline(
             If no framework is specified, will default to the one currently installed. If no framework is specified
             and both frameworks are installed, will default to the framework of the :obj:`model`, or to PyTorch if no
             model is provided.
+        use_fast (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            Whether or not to use a Fast tokenizer if possible (a :obj:`~transformers.PreTrainedTokenizerFast`).
```
Comment (sgugger) [inline src/transformers/pipelines.py:2752]:
```suggestion
            Whether or not to use a Fast tokenizer if possible (a :class:`~transformers.PreTrainedTokenizerFast`).
```

**Review by sgugger — APPROVED** [review]
Great! I like the decoupling small/big models a lot. Tests look super clean!

**Comment by thomwolf** [comment]
Ok. great, thanks @sgugger and @sshleifer.

I took the occasion to speed up the CI test suite (cc @stas00) by:
- spinning out the pipeline test in a separate job
- reducing the tf+pt tests overhead by focusing them on the PT+TF cross interaction (adding new tests) and removing the double testing with the tf and pytorch standalone tests.

Ready to merge imo.

**Review by LysandreJik — COMMENTED** [review]
Tests LGTM as seen offline.

We only need to speedup the TF tests (cf @stas00 :smile:) and we'll have a blazing fast test suite!

**Review by LysandreJik — APPROVED** [review]

File: src/transformers/testing_utils.py [hunk 3]
```diff
@@ -59,10 +59,50 @@ def parse_int_from_env(key, default=None):
...
+_run_pipeline_tests = parse_flag_from_env("RUN_PIPELINE_TESTS", default=False)
 _tf_gpu_memory_limit = parse_int_from_env("TF_GPU_MEMORY_LIMIT", default=None)
 
 
+def is_pt_tf_cross_test(test_case):
+    """
+    Decorator marking a test as a test that control interactions between PyTorch and TensorFlow.
+
+    PT+TF tests are skipped by default and we can run only them by setting RUN_PT_TF_CROSS_TESTS environment variable
+    to a truthy value and selecting the is_pipeline_test pytest mark.
```
Comment (sgugger) [inline src/transformers/testing_utils.py:73]:
```suggestion
    to a truthy value and selecting the is_pt_tf_cross_test pytest mark.
```

**Review by sgugger — APPROVED** [review]
Love the new blazing fast cross tests :-)

**Comment by stas00** [comment]
Whoah! this is amazing - the slowest job is now the fastest! Thank you!!!

I think there is only one potential issue with it - the half-baked until now codecov report is now completely useless since it no longer covers all tests so just as well remove it completely.

**Comment by LysandreJik** [comment]
@stas00 Did just that https://github.com/huggingface/transformers/commit/829b9f8cc321aa28396e6203e0f21eed26b132f7

Removed codecov from the repo as well.

**Comment by stas00** [comment]
It's still there ;)
```
.circleci/config.yml:            - run: pip install codecov pytest-cov
.circleci/config.yml:            - run: RUN_PT_TF_CROSS_TESTS=1 python -m pytest -n 8 --dist=loadfile -rA -s ./tests/ -m is_pt_tf_cross_test --cov --durations=0 | tee output.txt
.circleci/config.yml:            - run: codecov
```

**Pull request merged**